### PR TITLE
Fix '%z' and '%Z in '%{...}' expando

### LINKED ISSUE #4335: Timezone broken in index-format-hook display
## Expected Behaviour

I have the following `index-format-hook`:

```
index-format-hook  date  "~d<1d"    "%{%H:%M%z}"
```

When a message with timestamp e.g. "08:46:06 +0200" arrives, I expect that to be rendered as "08:46+0200".

## Actual Behaviour

`%z` always seems to expand to "+0000", yielding "08:46+0000", which is wrong. `%Z` expands to `GMT` in this case.

## How often does this happen?

- When I upgraded from 20231103 to 20240425. It worked before.

## NeoMutt Version

```
NeoMutt 20240425
Copyright (C) 2015-2024 Richard Russon and friends
NeoMutt comes with ABSOLUTELY NO WARRANTY; for details type 'neomutt -vv'.
NeoMutt is free software, and you are welcome to redistribute it
under certain conditions; type 'neomutt -vv' for details.

System: Linux 6.8.9-amd64 (x86_64)
ncurses: ncurses 6.5.20240427 (compiled with 6.5.20240427)
libidn2: 2.3.7 (compiled with 2.3.7)
GPGME: 1.18.0
GnuTLS: 3.8.5
libnotmuch: 5.6.0
PCRE2: 10.42 2022-12-11
storage: tokyocabinet, lmdb
compression: lz4, zlib, zstd

Configure options: --build=x86_64-linux-gnu --prefix=/usr {--includedir=${prefix}/include} {--mandir=${prefix}/share/man} {--infodir=${prefix}/share/info} --sysconfdir=/etc --localstatedir=/var --disable-option-checking --disable-silent-rules {--libdir=${prefix}/lib/x86_64-linux-gnu} --runstatedir=/run --disable-maintainer-mode --disable-dependency-tracking --mandir=/usr/share/man --libexecdir=/usr/libexec --with-mailpath=/var/mail --autocrypt --disable-paths-in-cflags --full-doc --gnutls --gpgme --gsasl --gss --lmdb --lua --lz4 --notmuch --mixmaster --pcre2 --sqlite --tokyocabinet --zlib --zstd

Compilation CFLAGS: -g -O2 -Werror=implicit-function-declaration -fstack-protector-strong -fstack-clash-protection -Wformat -Werror=format-security -fcf-protection -fno-delete-null-pointer-checks -D_ALL_SOURCE=1 -D_GNU_SOURCE=1 -D__EXTENSIONS__ -D_XOPEN_SOURCE_EXTENDED -DNCURSES_WIDECHAR -isystem /usr/include/mit-krb5 -O2

Compile options:
  +autocrypt +fcntl -flock -fmemopen +futimens +getaddrinfo +gnutls +gpgme 
  +gsasl +gss +hcache -homespool +idn +inotify -locales_hack +lua +mixmaster 
  +nls +notmuch -openssl +pcre2 +pgp -sasl +smime +sqlite +truecolor 

MAILPATH="/var/mail"
MIXMASTER="mixmaster"
PKGDATADIR="/usr/share/neomutt"
SENDMAIL="/usr/sbin/sendmail"
SYSCONFDIR="/etc"
```

## Extra Info

* Operating System and its version: Debian unstable
* Were you using multiple copies of NeoMutt at once? Multiple
* Were you using 'screen' or 'tmux'? No
* Is your email local (maildir) or remote (IMAP)? Local

### PR DESCRIPTION
**Pull request by gahr**
Fixes #4335

See https://github.com/neomutt/neomutt/pull/4339#issuecomment-2202293925 for an explanation of what I think is going on here.

### CODE REVIEW

File: hdrline.c [hunk 1]
```diff
@@ -461,6 +459,13 @@ void index_date(const struct ExpandoNode *node, void *data,
   ASSERT(len < sizeof(tmp2));
   mutt_strn_copy(tmp2, start, len, sizeof(tmp2));
 
+  // The sender's time zone might only be available as a numerical offset, so "%Z" behaves like "%z".
+  char *bigz = strstr(tmp2, "%Z");
```
Comment (g0mb4) [inline hdrline.c:463]: We should handle the case when multiple `%Z`s are present (to be fully compatible with the old parser)

**Review by g0mb4 — CHANGES_REQUESTED** [review]
I like this solution better. The results are compared with v20230517, and they are the same (except for multiple `%Z`s).

**Comment by g0mb4** [comment]
Ohh, `$index_format`'s`%d` (`index_d`) can be buggy if `$date_format` contains a `%z` or `%Z`.  It is the same problem.

**Comment by gahr** [comment]
> Ohh, `$index_format`'s`%d` (`index_d`) can be buggy if `$date_format` contains a `%z` or `%Z`. It is the same problem.

Yeah - most of the code is duplicated. Let me refactor it.

**Comment by g0mb4** [comment]
The output is correct, tested with:

```
color index_date yellow blue
set date_format = "%H:%M %z %Z %Z"
set index_format = "%(%H:%M) - %[%H:%M] - %{%H:%M %z %Z %Z} - %d"
```

but  it leaks.

```
==164185==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 150 byte(s) in 10 object(s) allocated from:
    #0 0x7a4a3f0b4887 in __interceptor_malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:145
    #1 0x5ad9d158e836 in mutt_mem_malloc mutt/memory.c:96
    #2 0x5ad9d159f16c in mutt_strn_dup mutt/string.c:385
    #3 0x5ad9d111f549 in index_email_date /home/gmb/prog/neomutt/neomutt/hdrline.c:355
    #4 0x5ad9d1121844 in index_d /home/gmb/prog/neomutt/neomutt/hdrline.c:692
    #5 0x5ad9d14e4c66 in node_expando_render expando/node_expando.c:355
    #6 0x5ad9d14e923a in node_render expando/render.c:55
    #7 0x5ad9d14ddfe0 in expando_render expando/expando.c:119
    #8 0x5ad9d14defa5 in expando_filter expando/filter.c:154
    #9 0x5ad9d1129a20 in mutt_make_string /home/gmb/prog/neomutt/neomutt/hdrline.c:1748
    #10 0x5ad9d11a82f6 in index_make_entry index/dlg_index.c:891
    #11 0x5ad9d135d14a in menu_redraw_index menu/draw.c:338
    #12 0x5ad9d11c8ecd in index_repaint index/index.c:629
    #13 0x5ad9d1307a38 in window_repaint gui/mutt_window.c:598
    #14 0x5ad9d1307aea in window_repaint gui/mutt_window.c:604
    #15 0x5ad9d1307aea in window_repaint gui/mutt_window.c:604
    #16 0x5ad9d1307aea in window_repaint gui/mutt_window.c:604
    #17 0x5ad9d1307aea in window_repaint gui/mutt_window.c:604
    #18 0x5ad9d1307aea in window_repaint gui/mutt_window.c:604
    #19 0x5ad9d1307d05 in window_redraw gui/mutt_window.c:642
    #20 0x5ad9d11ab9d7 in dlg_index index/dlg_index.c:1242
    #21 0x5ad9d114cdc1 in main /home/gmb/prog/neomutt/neomutt/main.c:1432
    #22 0x7a4a3ea29d8f in __libc_start_call_main ../sysdeps/nptl/libc_start_call_main.h:58

Direct leak of 150 byte(s) in 10 object(s) allocated from:
    #0 0x7a4a3f0b4887 in __interceptor_malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:145
    #1 0x5ad9d158e836 in mutt_mem_malloc mutt/memory.c:96
    #2 0x5ad9d159f16c in mutt_strn_dup mutt/string.c:385
    #3 0x5ad9d111f549 in index_email_date /home/gmb/prog/neomutt/neomutt/hdrline.c:355
    #4 0x5ad9d111fda8 in index_date /home/gmb/prog/neomutt/neomutt/hdrline.c:462
    #5 0x5ad9d14e4c66 in node_expando_render expando/node_expando.c:355
    #6 0x5ad9d14e923a in node_render expando/render.c:55
    #7 0x5ad9d14ddfe0 in expando_render expando/expando.c:119
    #8 0x5ad9d14defa5 in expando_filter expando/filter.c:154
    #9 0x5ad9d1129a20 in mutt_make_string /home/gmb/prog/neomutt/neomutt/hdrline.c:1748
    #10 0x5ad9d11a82f6 in index_make_entry index/dlg_index.c:891
    #11 0x5ad9d135d14a in menu_redraw_index menu/draw.c:338
    #12 0x5ad9d11c8ecd in index_repaint index/index.c:629
    #13 0x5ad9d1307a38 in window_repaint gui/mutt_window.c:598
    #14 0x5ad9d1307aea in window_repaint gui/mutt_window.c:604
    #15 0x5ad9d1307aea in window_repaint gui/mutt_window.c:604
    #16 0x5ad9d1307aea in window_repaint gui/mutt_window.c:604
    #17 0x5ad9d1307aea in window_repaint gui/mutt_window.c:604
    #18 0x5ad9d1307aea in window_repaint gui/mutt_window.c:604
    #19 0x5ad9d1307d05 in window_redraw gui/mutt_window.c:642
    #20 0x5ad9d11ab9d7 in dlg_index index/dlg_index.c:1242
    #21 0x5ad9d114cdc1 in main /home/gmb/prog/neomutt/neomutt/main.c:1432
    #22 0x7a4a3ea29d8f in __libc_start_call_main ../sysdeps/nptl/libc_start_call_main.h:58

Direct leak of 60 byte(s) in 10 object(s) allocated from:
    #0 0x7a4a3f0b4887 in __interceptor_malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:145
    #1 0x5ad9d158e836 in mutt_mem_malloc mutt/memory.c:96
    #2 0x5ad9d159f16c in mutt_strn_dup mutt/string.c:385
    #3 0x5ad9d111f549 in index_email_date /home/gmb/prog/neomutt/neomutt/hdrline.c:355
    #4 0x5ad9d111fbb1 in index_date_local /home/gmb/prog/neomutt/neomutt/hdrline.c:432
    #5 0x5ad9d14e4c66 in node_expando_render expando/node_expando.c:355
    #6 0x5ad9d14e923a in node_render expando/render.c:55
    #7 0x5ad9d14ddfe0 in expando_render expando/expando.c:119
    #8 0x5ad9d14defa5 in expando_filter expando/filter.c:154
    #9 0x5ad9d1129a20 in mutt_make_string /home/gmb/prog/neomutt/neomutt/hdrline.c:1748
    #10 0x5ad9d11a82f6 in index_make_entry index/dlg_index.c:891
    #11 0x5ad9d135d14a in menu_redraw_index menu/draw.c:338
    #12 0x5ad9d11c8ecd in index_repaint index/index.c:629
    #13 0x5ad9d1307a38 in window_repaint gui/mutt_window.c:598
    #14 0x5ad9d1307aea in window_repaint gui/mutt_window.c:604
    #15 0x5ad9d1307aea in window_repaint gui/mutt_window.c:604
    #16 0x5ad9d1307aea in window_repaint gui/mutt_window.c:604
    #17 0x5ad9d1307aea in window_repaint gui/mutt_window.c:604
    #18 0x5ad9d1307aea in window_repaint gui/mutt_window.c:604
    #19 0x5ad9d1307d05 in window_redraw gui/mutt_window.c:642
    #20 0x5ad9d11ab9d7 in dlg_index index/dlg_index.c:1242
    #21 0x5ad9d114cdc1 in main /home/gmb/prog/neomutt/neomutt/main.c:1432
    #22 0x7a4a3ea29d8f in __libc_start_call_main ../sysdeps/nptl/libc_start_call_main.h:58

Direct leak of 60 byte(s) in 10 object(s) allocated from:
    #0 0x7a4a3f0b4887 in __interceptor_malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:145
    #1 0x5ad9d158e836 in mutt_mem_malloc mutt/memory.c:96
    #2 0x5ad9d159f16c in mutt_strn_dup mutt/string.c:385
    #3 0x5ad9d111f549 in index_email_date /home/gmb/prog/neomutt/neomutt/hdrline.c:355
    #4 0x5ad9d111f9ba in index_date_recv_local /home/gmb/prog/neomutt/neomutt/hdrline.c:402
    #5 0x5ad9d14e4c66 in node_expando_render expando/node_expando.c:355
    #6 0x5ad9d14e923a in node_render expando/render.c:55
    #7 0x5ad9d14ddfe0 in expando_render expando/expando.c:119
    #8 0x5ad9d14defa5 in expando_filter expando/filter.c:154
    #9 0x5ad9d1129a20 in mutt_make_string /home/gmb/prog/neomutt/neomutt/hdrline.c:1748
    #10 0x5ad9d11a82f6 in index_make_entry index/dlg_index.c:891
    #11 0x5ad9d135d14a in menu_redraw_index menu/draw.c:338
    #12 0x5ad9d11c8ecd in index_repaint index/index.c:629
    #13 0x5ad9d1307a38 in window_repaint gui/mutt_window.c:598
    #14 0x5ad9d1307aea in window_repaint gui/mutt_window.c:604
    #15 0x5ad9d1307aea in window_repaint gui/mutt_window.c:604
    #16 0x5ad9d1307aea in window_repaint gui/mutt_window.c:604
    #17 0x5ad9d1307aea in window_repaint gui/mutt_window.c:604
    #18 0x5ad9d1307aea in window_repaint gui/mutt_window.c:604
    #19 0x5ad9d1307d05 in window_redraw gui/mutt_window.c:642
    #20 0x5ad9d11ab9d7 in dlg_index index/dlg_index.c:1242
    #21 0x5ad9d114cdc1 in main /home/gmb/prog/neomutt/neomutt/main.c:1432
    #22 0x7a4a3ea29d8f in __libc_start_call_main ../sysdeps/nptl/libc_start_call_main.h:58

SUMMARY: AddressSanitizer: 420 byte(s) leaked in 40 allocation(s).
```

**Comment by gahr** [comment]
ah sure, my bad - forgot to free the temporary.

**Comment by g0mb4** [comment]
Works.

**Review by g0mb4 — APPROVED** [review]

**Comment by gahr** [comment]
> Can we add a test for this?

Good question, I have very limited familiarity with the expando machinery. @g0mb4 how hard would it be to add a unit test for this?

**Comment by g0mb4** [comment]
Well, I have to refresh my memory on the tests (I created them to test the expando library, not a specific expando that Neomutt uses), but I think it is possible. I can create the test in a couple of days if that is okay.

**Comment by g0mb4** [comment]
I'm afraid that there is no easy way to test this. I see two options:

1.  Copy/Paste the expando callback and test that. This way, we have to manually sync the callbacks.
2.  Use the callback NeoMutt uses.  This would require linking a lot of the `.o` files from NeoMutt.... (there are a *LOT* of stubs to prevent that)

Any opinions?

**Comment by g0mb4** [comment]
3. Wilder idea: Run neomutt with a specific `.rc` file against a test mailbox, redirect the output into a file, and compare it against a golden file.

**Comment by roccoblues** [comment]
I'm probably missing something but why can't we write unit tests for the `index_email_date` function? Is that what you mean with: 

>     2. Use the callback NeoMutt uses.  This would require linking a lot of the `.o` files from NeoMutt.... (there are a _LOT_ of stubs to prevent that)

?

**Comment by g0mb4** [comment]
Yep, but if we can test one callback, we are able to test all of them, but the required information on how to parse/render an expando is not part of `libexpando`. If we truly want to test them we have to use the same parameters as NeoMutt.

**Comment by gahr** [comment]
Alright - I'd rather merge this anyway: @roccoblues ack?

**Review by roccoblues — APPROVED** [review]

**Pull request merged**